### PR TITLE
Possibility to separately specify log levels for log handlers - basic

### LINKED ISSUE #32619: Possibility to separately specify log levels for log handlers
### Description

At this moment, we support three log handlers: `console`, `file`, and `syslog`. However, we are not able to specify what log records should be included for each particular log handler. 

Currently, we have `log-level` property, which provides the ability to set the root level for all log handlers, or specifying log levels for particular categories (`org.keycloak:debug`, `io.quarkus:trace`,...).

We should have more fine-grained settings for the log handlers to meet the requirements of distinct deployments. 

The log levels might be controlled via properties associated with each log handler, such as:

- `log-console-level` - Console
- `log-file-level` - File
- `log-syslog-level` - Syslog

We should be able to meet, at least, this behavior (for Syslog, but for File would be the same):

- `--log-level=info --log-syslog-level=debug` => console shows **INFO**, Syslog shows **DEBUG**
- `--log-level=debug --log-syslog-level=info` => console shows **DEBUG**, Syslog shows **INFO**
- (will be added more later)

We could go even further and have the specific log handlers levels even more fine-grained. Specifically, be able to set log levels for particular categories for a specific log handler => `log-syslog-level=debug,org.keycloak.events:trace`, or something like that.

### Details

Quarkus provides the ability to manage each log handlers levels separately, but it's a little bit tricky with the root logger levels. It means that the root logger creates some sort of a minimum bar, that each logger accepts and even when you set TRACE for Syslog and the root log level is INFO, only the INFO log records will be displayed for Syslog. This should be solved and be user-friendly enough.

### PR DESCRIPTION
**Pull request by mabartos**
- Closes #32619

Descoped version of https://github.com/keycloak/keycloak/pull/32620, with slightly different logic around log levels - based on the current JBoss log manager/Quarkus implementation. 

The logic around log levels for handlers might not be very intuitive, and the docs in the Quarkus [Logging configuration guide](https://quarkus.io/guides/logging) do not touch more details as well.

### CODE REVIEW

**Comment by mabartos** [comment]
@keycloak/cloud-native-maintainers Hey guys, could someone please check it? Thanks.

@dmlloyd May I ask you for a brief review of the 'General principle' section in docs? Screenshots are attached in the PR description. Thank you!

File: docs/guides/server/logging.adoc [hunk 1]
```diff
@@ -71,15 +74,90 @@ To enable log handlers, enter the following command:
...
+
+.Root log level `info`
+[#img-log-level-info]
+image::logging/log-levels-root-info.png[Root log level info,720]
+
+As you can see in the figure <<img-log-level-info>>, the root log level, specified via the `log-level` property, sets the minimal level for log records.
+Therefore, even when their log level is the finest, it is ignored and respects the parent level.
+
+The _Min Level_ bar represents the minimal allowed log level in the application (`trace` by default).
+Log records with finer log levels will never be present in {project_name}.
```
Comment (vmuzikar) [inline docs/guides/server/logging.adoc:127]: I'm not sure what this sentence means.

File: docs/guides/server/logging.adoc [hunk 2]
```diff
@@ -71,15 +74,90 @@ To enable log handlers, enter the following command:
...
+It is controlled by the Quarkus property `quarkus.log.min-level`.
```
Comment (vmuzikar) [inline docs/guides/server/logging.adoc:128]: I would not mention Quarkus options here. User cares about Keycloak and not care about implementational details.

File: docs/guides/server/logging.adoc [hunk 3]
```diff
@@ -71,15 +74,90 @@ To enable log handlers, enter the following command:
...
+Only the `log-level` property is changed, and other log handlers are set as follows:
+
+<@kc.start parameters="--log=syslog,console,file --log-level=info --log-syslog-level=all --log-console-level=debug --log-file-level=trace"/>
+
+.Root log level `info`
+[#img-log-level-info]
+image::logging/log-levels-root-info.png[Root log level info,720]
+
+As you can see in the figure <<img-log-level-info>>, the root log level, specified via the `log-level` property, sets the minimal level for log records.
+Therefore, even when their log level is the finest, it is ignored and respects the parent level.
```
Comment (vmuzikar) [inline docs/guides/server/logging.adoc:124]:
```suggestion
Therefore, even when their log level is more verbose than the root logger, it is ignored and respects the root one.
```

File: docs/guides/server/logging.adoc [hunk 4]
```diff
@@ -71,15 +74,90 @@ To enable log handlers, enter the following command:
...
+image::logging/log-levels-root-trace.png[Root log level trace,720]
+
+In this case, by changing the `log-level` property to `trace` level, visible in the figure <<img-log-level-trace>>, log handlers are not restricted anymore.
+
+.Additional categories
+[#img-log-level-categories]
+image::logging/log-levels-root-info-categories.png[Additional categories,720]
+
+As mentioned in the <<Configuring category-specific log levels>> section, some additional categories can be set.
+In figure <<img-log-level-categories>>, the `org.keycloak.events` logger uses the `debug` log level, where its log records will be part of the output next to the `info` log records.
```
Comment (vmuzikar) [inline docs/guides/server/logging.adoc:141]: This feels a bit confusing and IMHO is making the impression we support setting categories on log handler level. I'd rephrase the whole paragraph a bit.

File: docs/guides/server/logging.adoc [hunk 5]
```diff
@@ -71,15 +74,90 @@ To enable log handlers, enter the following command:
...
+.Example: `debug` for file handler, but `info` for console handler:
+<@kc.start parameters="--log=console,file --log-level=debug --log-console-level=info"/>
+
+The root log level is set to `debug`, so every log handler inherits the value - so does the file log handler.
+To hide `debug` records in the console, we need to set the minimal bar `info` level for the console handler.
+
+.Example: `warn` for all handlers, but `debug` for file handler:
+<@kc.start parameters="--log=console,file,syslog --log-level=debug --log-console-level=warn --log-syslog-level=warn"/>
+
+The root level must be set to the finest required level (`debug` in this case), and other log handlers must be amended accordingly.
```
Comment (vmuzikar) [inline docs/guides/server/logging.adoc:154]:
```suggestion
The root level must be set to the most verbose required level (`debug` in this case), and other log handlers must be amended accordingly.
```

File: docs/guides/server/logging.adoc [hunk 6]
```diff
@@ -71,15 +74,90 @@ To enable log handlers, enter the following command:
...
+It is necessary to understand that setting the log levels for each particular handler *does not override the root level* specified in the `log-level` property.
+It specifies the minimal allowed level for logs shown in the output.
+It respects levels specified in the `log-level` property, which creates a minimal log level bar for log records.
+
+Specifically, when an arbitrary log level is specified for the handler, it does not mean the log records with the log level will be present in the output.
+The root log level needs to be assessed as well, as it determines the minimal level for exposed log records.
+
+Log handler levels provide the *restriction for the root log level*, so the default value is `all` - without any restriction.
+
+==== Visualization
```
Comment (vmuzikar) [inline docs/guides/server/logging.adoc:112]: I was thinking if we even need the visualization. It'd be a bit hard to maintain if changes are needed (without source files). And the examples (if we move the written explanation from here to there) are good on its own, IMHO. @mabartos WDYT?

File: quarkus/tests/integration/src/test/java/org/keycloak/it/cli/dist/QuarkusPropertiesAutoBuildDistTest.java [hunk 7]
```diff
@@ -39,38 +36,10 @@
 @TestMethodOrder(OrderAnnotation.class)
 public class QuarkusPropertiesAutoBuildDistTest {
 
-    @Test
```
Comment (vmuzikar) [inline quarkus/tests/integration/src/test/java/org/keycloak/it/cli/dist/QuarkusPropertiesAutoBuildDistTest.java:42]: I might be missing something but why are we removing so many tests?

**Review by vmuzikar — CHANGES_REQUESTED** [review]
@mabartos Nice work, thank you for the PR! Added some comments.

File: quarkus/tests/integration/src/test/java/org/keycloak/it/cli/dist/QuarkusPropertiesAutoBuildDistTest.java (hunk 7 above)
Comment (mabartos) [reply quarkus/tests/integration/src/test/java/org/keycloak/it/cli/dist/QuarkusPropertiesAutoBuildDistTest.java:42]: @vmuzikar 
All these tests are not relevant anymore as they directly touch the `quarkus.log.console.level` prop, which was not exposed before. In this case, we map to the Qks property from our KC props `--log-console-level`. Even when the Qks prop is set, it's overridden by our property mapper - by a specified value or default value.

"They're testing setting Qks props, should we replace them with sth else?" - IMO, we do not need so many tests around these configs as even before, they did not cover different behavior than the ones that stay (hibernate and datasources).

So, I'd say it's ok to have it like this.
Is it ok?

Comment (vmuzikar) [reply quarkus/tests/integration/src/test/java/org/keycloak/it/cli/dist/QuarkusPropertiesAutoBuildDistTest.java:42]: As far as I can tell, the removed tests were not focused on `quarkus.log.console.level` but on other things like if reaug acts as it's supposed to (`reAugOnFirstRun`, `testQuarkusRuntimePropDoesNotTriggerReAug`, `testNoReAugAfterChangingRuntimeProperty`, ...).

File: docs/guides/server/logging.adoc (hunk 6 above)
Comment (mabartos) [reply docs/guides/server/logging.adoc:112]: I'm not a big fan of comprehensive docs essays when a simple visualization might tell you more. In this case, fully understanding the log levels for handlers and for the root is crucial.

I'd like to prevent creating multiple issues like: "Hey, I set `debug` for syslog, but it shows only `info`". Even when we have it documented, I like to have it as clear as possible. 

Yes, I've been thinking the same about maintainability, but I've convinced myself to have it there. The logic that is visualized is the standard for Quarkus logging and will not change drastically - might be extended, but not changed (AFAIK). That's why I also pinged @dmlloyd for more information. IMHO, the visualization brings more clearance on the log levels and can bring some value to users, at least until the time when sth is changed. 

I expect that the relevance will last a big portion of the time, and it can be removed when it's not relevant anymore.

@vmuzikar WDYT?

File: quarkus/tests/integration/src/test/java/org/keycloak/it/cli/dist/QuarkusPropertiesAutoBuildDistTest.java (hunk 7 above)
Comment (mabartos) [reply quarkus/tests/integration/src/test/java/org/keycloak/it/cli/dist/QuarkusPropertiesAutoBuildDistTest.java:42]: @vmuzikar You're right, I'll revisit that.

File: docs/guides/server/logging.adoc [hunk 8]
```diff
@@ -71,15 +74,90 @@ To enable log handlers, enter the following command:
...
+Accepted log levels are specified in <<Log levels>> section, and *must be in lowercase*.
+
+=== General principle
+
+It is necessary to understand that setting the log levels for each particular handler *does not override the root level* specified in the `log-level` property.
+It specifies the minimal allowed level for logs shown in the output.
+It respects levels specified in the `log-level` property, which creates a minimal log level bar for log records.
+
+Specifically, when an arbitrary log level is specified for the handler, it does not mean the log records with the log level will be present in the output.
+The root log level needs to be assessed as well, as it determines the minimal level for exposed log records.
```
[the remark this answers is not in the document]
Comment (mabartos) [reply docs/guides/server/logging.adoc]: Even though I accepted your suggestion, I thought about it once again, and I think the first idea was better. I understand it might be confusing a bit, but it depends on the PoV - verbosity x severity levels. 

For multiple(all I know) log levels representations, the levels are sorted descending based on the severity. All docs show the table of log levels also based on the severity. So, IMHO, it's more intuitive to say the `trace` is minimal instead of `fatal` is minimal - so, **severity**, no verbosity. 

java.util.logging.Level:
```
SEVERE = 1000
WARNING = 900
....
FINEST = 300
``` 

JBoss LogManager:
```
FATAL = 1100
ERROR = 1000
TRACE = 400
```

Also the Quarkus sets the minimal level (`quarkus.log.min-level`) to `debug` (based on severity).

So I'd propose using the 'minimal' allowed level as it's more natural for users imagine the log level table, where the minimal item is `TRACE`/`FINEST`. 

@vmuzikar Is it ok? Or should I make it more clear - explicitly mention the terminology?

[the remark this answers is not in the document]
Comment (vmuzikar) [reply docs/guides/server/logging.adoc]: That's a good point we're basically aligning with Quarkus on the wording but IMHO it does not necessarily mean it's correct. :) What I find confusing is that AFAIK there's an established term "increase log level" but at the same time we're saying e.g. `debug` is less than `error`.

But maybe we could find some middle ground here to make it clear. Instead of minimal/maximal we'd either say most/least verbose, or most/least severe. WDYT?

File: docs/guides/server/logging.adoc (hunk 6 above)
Comment (vmuzikar) [reply docs/guides/server/logging.adoc:112]: It's not a blocker for me to keep the visualization. I just think the concrete examples below the images made it even more clear than the images itself. And that's a not a lot of text. At least I felt this way. :)

File: quarkus/tests/integration/src/test/java/org/keycloak/it/cli/dist/QuarkusPropertiesAutoBuildDistTest.java (hunk 7 above)
Comment (mabartos) [reply quarkus/tests/integration/src/test/java/org/keycloak/it/cli/dist/QuarkusPropertiesAutoBuildDistTest.java]: Fixed.

File: docs/guides/server/logging.adoc (hunk 8 above)
[the remark this answers is not in the document]
Comment (mabartos) [reply docs/guides/server/logging.adoc:109]: Fair enough; I'll do. Thanks!

**Comment by dmlloyd** [comment]
The explanations all seem correct to me.

File: docs/guides/server/logging.adoc (hunk 6 above)
Comment (mabartos) [reply docs/guides/server/logging.adoc:113]: @vmuzikar It depends on our consensus here, I'm not declining to remove it :) I just think it brings some added value as the log-level is more understandable, even for some special cases - it helps to imagine the principle better. As even Quarkus does not provide a clear explanation, I like it this way.

But if it can be found disruptive, I can remove it right away.

File: docs/guides/server/logging.adoc (hunk 8 above)
[the remark this answers is not in the document]
Comment (mabartos) [reply docs/guides/server/logging.adoc]: @vmuzikar In the end, I used some kind of hybrid approach to properly reference the log-level bars. I hope it's okay. Please review.

**Review by vmuzikar — COMMENTED** [review]
@mabartos Thank you for the update! Added some last nitpick, other than that LGTM.

**Comment by mabartos** [comment]
@vmuzikar Rebased, resolved conflicts, and applied the last changes.

**Review by vmuzikar — DISMISSED** [review]

**Comment by mabartos** [comment]
@vmuzikar Also options description might be edited. I'll provide it.

**Comment by mabartos** [comment]
@vmuzikar Should be complete.

**Review by vmuzikar — APPROVED** [review]
LGTM, thanks.
